Std/Tactic/NormCast.lean

```diff
@@ -0,0 +1,333 @@
+def normCastTarget : TacticM Unit :=
+  liftMetaTactic1 fun goal => do
+    let tgt ← instantiateMVars (← goal.getType)
+    let prf ← derive tgt
+    applySimpResultToTarget goal tgt prf
+
+/-- Implementation of the `norm_cast` tactic when operating on a hypothesis. -/
+def normCastHyp (fvarId : FVarId) : TacticM Unit :=
+  liftMetaTactic1 fun goal => do
+    let hyp ← instantiateMVars (← fvarId.getDecl).type
+    let prf ← derive hyp
+    return (← applySimpResultToLocalDecl goal fvarId prf false).map (·.snd)
+
+/-- Implementation of `norm_cast` (the full `norm_cast` calls `trivial` afterwards). -/
+elab "norm_cast0" loc:((location)?) : tactic =>
+  withMainContext do
+    match expandOptLocation loc with
+    | Location.targets hyps target =>
+      if target then normCastTarget
+      (← getFVarIds hyps).forM normCastHyp
+    | Location.wildcard =>
+      normCastTarget
+      (← (← getMainGoal).getNondepPropHyps).forM normCastHyp
+
+/-- `assumption_mod_cast` runs `norm_cast` on the goal. For each local hypothesis `h`, it also
+normalizes `h` and tries to use that to close the goal. -/
+macro "assumption_mod_cast" : tactic => `(tactic| norm_cast0 at * <;> assumption)
+
+/--
+The `norm_cast` family of tactics is used to normalize casts inside expressions.
+It is basically a simp tactic with a specific set of lemmas to move casts
+upwards in the expression.
+Therefore it can be used more safely as a non-terminating tactic.
+It also has special handling of numerals.
+
+For instance, given an assumption
+```lean
+a b : ℤ
+h : ↑a + ↑b < (10 : ℚ)
+```
+
+writing `norm_cast at h` will turn `h` into
+```lean
+h : a + b < 10
+```
+
+You can also use `exact_mod_cast`, `apply_mod_cast`, `rw_mod_cast`
+or `assumption_mod_cast`.
+Writing `exact_mod_cast h` and `apply_mod_cast h` will normalize the goal and
+`h` before using `exact h` or `apply h`.
+Writing `assumption_mod_cast` will normalize the goal and for every
+expression `h` in the context it will try to normalize `h` and use
+`exact h`.
+`rw_mod_cast` acts like the `rw` tactic but it applies `norm_cast` between steps.
+
+See also `push_cast`, for move casts inwards.
+
+The implementation and behavior of the `norm_cast` family is described in detail at
+<https://lean-forward.github.io/norm_cast/norm_cast.pdf>.
+-/
+macro "norm_cast" loc:(location)? : tactic =>
+  `(tactic| norm_cast0 $[$loc]? <;> try trivial)
+
+/--
+Rewrite with the given rules and normalize casts between steps.
+-/
+syntax "rw_mod_cast" (config)? rwRuleSeq (location)? : tactic
+macro_rules
+  | `(tactic| rw_mod_cast $[$config]? [$rules,*] $[$loc]?) => do
+    let tacs ← rules.getElems.mapM fun rule =>
+      `(tactic| (norm_cast at *; rw $[$config]? [$rule] $[$loc]?))
+    `(tactic| ($[$tacs]*))
+
+/--
+Normalize the goal and the given expression, then close the goal with exact.
+-/
+macro "exact_mod_cast " e:term : tactic => `(tactic| exact mod_cast ($e : _))
```

Comment:
Is the empty type ascription needed here? Besides the part where this annotation does not work in lean 4 (which is why `(e :)` was added), if we can get away with just `exact mod_cast $e` then perhaps we don't need `exact_mod_cast` at all as it's just syntax for something more compositional with the same number of characters.

Reply:
Unfortunately a small fraction of the `exact_mod_cast`s in Mathlib can not be replaced by `exact mod_cast`. See #8222.